File: 748 - Shortest Completing Word.js
/**
 * @param {string} licensePlate
 * @param {string[]} words
 * @return {string}
 */
let m = new Map();
var shortestCompletingWord = function (licensePlate, words) {
    let res = 'asdjioioajsdadsjioasdj';
    let patt = /[a-zA-Z]/;
    m.clear();
    for (let c of licensePlate) {
        if (patt.test(c)) {
            c = c.toLowerCase();
            if (m.has(c)) {
                m.set(c, m.get(c) + 1);
            } else {
                m.set(c, 1);
            }
        }
    }

    for (let word of words) {
        if (check(word)) {
            res = word.length < res.length ? word : res;
        }
    }
    return res;
};

function check(word) {
    for (let c of m.keys()) {
        let p = new RegExp(c, "g");
        let arr = word.match(p);
        let count = arr == null ? 0 : arr.length;
        if (count >= m.get(c)) {
            continue;
        } else {
            return false;
        }
    }
    return true;
}

File: 748 - Shortest Completing Word.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    private Map<Character, Integer> m = new HashMap<>();

    public String shortestCompletingWord(String licensePlate, String[] words) {
        for (int i = 0; i < licensePlate.length(); i++) {
            char c = licensePlate.charAt(i);
            if (Character.isAlphabetic(c)) {
                c = Character.toLowerCase(c);
                m.put(c, (m.putIfAbsent(c, 0) == null ? 0 : m.get(c)) + 1);
            }
        }
        String res = "okokokokokokkokkokoko";
        for (String word : words) {
            if (check(word)) {
                res = word.length() < res.length() ? word : res;
            }
        }
        return res;
    }

    public boolean check(String word) {
        for (Character c : m.keySet()) {
            if (count(word, c) >= m.get(c)) {
                continue;
            } else {
                return false;
            }
        }
        return true;
    }

    public int count(String word, Character c) {
        int res = 0;
        for (int i = 0; i < word.length(); i++) {
            Character t = word.charAt(i);
            if (Character.toLowerCase(t) == c)
                res++;

        }
        return res;
    }

}